### PR TITLE
Fix #2024: TypeApply can be a final apply of a recursive method.

### LINKED ISSUE #2024: Tailrec label def failing compilation.
```scala
def assume[T]: Any = assume
```

fails to compile with the following error message 
```
value tailLabel2 does not take parameters
```

### PR DESCRIPTION
**Pull request by DarkDimius**
nullary methods with type parameters were handled wrongly, which led to
    absence of type application and a cryptic error message.

### CODE REVIEW

**Comment by DarkDimius** [comment]
@felixmulder, is our CI dead?

```
+ git init
Initialized empty Git repository in /drone/src/github.com/lampepfl/dotty/.git/
+ git remote add origin https://github.com/lampepfl/dotty.git
+ git fetch --no-tags origin +refs/pull/2025/merge:
fatal: Couldn't find remote ref refs/pull/2025/merge
exit status 128
```

**Comment by DarkDimius** [comment]
@nicolasstucki please review.

File: compiler/src/dotty/tools/backend/jvm/GenBCode.scala [hunk 1]
```diff
@@ -205,7 +205,15 @@ class GenBCodePipeline(val entryPoints: List[Symbol], val int: DottyBackendInter
         if (claszSymbol.isClass) // @DarkDimius is this test needed here?
           for (binary <- ctx.compilationUnit.pickled.get(claszSymbol.asClass)) {
             val dataAttr = new CustomAttr(nme.TASTYATTR.toString, binary)
-            (if (mirrorC ne null) mirrorC else plainC).visitAttribute(dataAttr)
+            val store = if (mirrorC ne null) mirrorC else plainC
+            store.visitAttribute(dataAttr)
+            if (ctx.settings.emitTasty.value) {
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/backend/jvm/GenBCode.scala]: This does not seem to be relevant to this PR as `-XemitTasty` is never enabled.

**Comment by odersky** [comment]
Let's make it a -Y option?

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala [hunk 2]
```diff
@@ -44,6 +44,7 @@ class ScalaSettings extends Settings.SettingGroup {
   /** -X "Advanced" settings
    */
   val Xhelp = BooleanSetting("-X", "Print a synopsis of advanced options.")
+  val emitTasty = BooleanSetting("-XemitTasty", "Generate tasty in separate *.tasty file.")
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: You should probably make it a `-Y` option or remove it from this PR.

**Comment by nicolasstucki** [comment]
LGTM

**Pull request merged**